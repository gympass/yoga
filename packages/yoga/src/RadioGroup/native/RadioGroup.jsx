import React, { useState } from 'react';
import { View } from 'react-native';
import { string, func, oneOfType, number, bool, node } from 'prop-types';
import styled from 'styled-components';

import RadioGroupContext from '../RadioGroupContext';
import RadioButton from './Button/RadioButton';

const Group = styled.View(
  () => `
    align-self: center;
    position: relative;
    flex-direction: row;
  `,
);

const GrayLine = styled.View(
  ({
    theme: {
      yoga: {
        components: { radiogroup },
      },
    },
  }) => `
    position: absolute;
    height: 100%;
    border-radius: ${radiogroup.button.border.radius}px;
    border-width: ${radiogroup.button.border.width}px;
    border-color: ${radiogroup.button.border.color};
  `,
);

const RadioGroup = ({
  onChange = () => {},
  selectedValue = '',
  small = false,
  children,
  ...rest
}) => {
  const [groupSize, setGroupSize] = useState(0);

  const isButton = React.Children.map(children, child => child.type).every(
    child => child === RadioButton,
  );

  return (
    <RadioGroupContext.Provider
      value={{
        selectedValue,
        onChange,
        small,
        ...rest,
      }}
    >
      <View>
        <Group
          {...rest}
          onLayout={({
            nativeEvent: {
              layout: { width },
            },
          }) => setGroupSize(width)}
        >
          {Boolean(isButton && groupSize) && (
            <GrayLine style={{ width: groupSize }} />
          )}

          {children}
        </Group>
      </View>
    </RadioGroupContext.Provider>
  );
};

RadioGroup.propTypes = {
  onChange: func,
  selectedValue: oneOfType([string, number]),
  small: bool,
  children: node.isRequired,
};

export default RadioGroup;
