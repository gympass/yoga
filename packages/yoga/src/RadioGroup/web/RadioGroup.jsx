import React from 'react';
import { string, func, oneOfType, number, bool, node, shape } from 'prop-types';
import styled from 'styled-components';

import RadioGroupContext from '../RadioGroupContext';

const Group = styled.div.attrs({
  role: 'radiogroup',
})`
  display: flex;

  width: min-content;

  ${({ full }) => `
    ${full ? `width: 100%;` : ``}
  `}
`;

const RadioGroup = ({
  name = '',
  onChange = () => {},
  selectedValue = '',
  small = false,
  full = false,
  children = null,
  style = {},
  ...rest
}) => (
  <RadioGroupContext.Provider
    value={{
      selectedValue,
      name,
      onChange,
      small,
      ...rest,
    }}
  >
    <Group full={full} style={style} {...rest}>
      {children}
    </Group>
  </RadioGroupContext.Provider>
);

RadioGroup.propTypes = {
  name: string,
  onChange: func,
  selectedValue: oneOfType([string, number]),
  small: bool,
  full: bool,
  children: node,
  style: shape({}),
};

export default RadioGroup;
