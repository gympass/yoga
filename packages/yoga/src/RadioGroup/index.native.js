import { RadioGroup, RadioButton, Radio } from './native';

const ExportRadioGroup = Object.assign(RadioGroup, {
  Button: RadioButton,
  Radio,
});

export default ExportRadioGroup;
