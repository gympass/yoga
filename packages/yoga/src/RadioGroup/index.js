import { RadioGroup, RadioButton, Radio } from './web';

const ExportRadioGroup = Object.assign(RadioGroup, {
  Button: RadioButton,
  Radio,
});

export default ExportRadioGroup;
