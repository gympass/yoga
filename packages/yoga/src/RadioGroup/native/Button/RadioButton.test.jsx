import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';

import RadioGroup from '../..';
import ThemeProvider from '../../../Theme';
import RadioGroupContext from '../../RadioGroupContext';

describe('<RadioGroup />', () => {
  describe('<RadioGroup.Button />', () => {
    describe('Snapshots', () => {
      it('should match snapshot with default radio button', () => {
        const { container, toJSON } = render(
          <ThemeProvider>
            <RadioGroup.Button>Radio 1</RadioGroup.Button>
          </ThemeProvider>,
        );

        expect(toJSON(container)).toMatchSnapshot();
      });

      it('should match snapshot with small radio button', () => {
        const { container, toJSON } = render(
          <RadioGroupContext.Provider value={{ small: true }}>
            <ThemeProvider>
              <RadioGroup.Button>Radio 1</RadioGroup.Button>
            </ThemeProvider>
          </RadioGroupContext.Provider>,
        );

        expect(toJSON(container)).toMatchSnapshot();
      });
    });

    describe('Props', () => {
      it('checked should be the selectedValue provided from Context', () => {
        const { getAllByText } = render(
          <RadioGroupContext.Provider value={{ selectedValue: 'Radio 2' }}>
            <ThemeProvider>
              <RadioGroup.Button>Radio 1</RadioGroup.Button>
              <RadioGroup.Button>Radio 2</RadioGroup.Button>
            </ThemeProvider>
          </RadioGroupContext.Provider>,
        );

        const [radio, radio2] = getAllByText(/^Radio/);

        expect(radio.props.checked).toBe(false);
        expect(radio2.props.checked).toBe(true);
      });
    });

    describe('Events', () => {
      it('should call onchange from Context', () => {
        const onChangeMock = jest.fn();
        const { getByText } = render(
          <RadioGroupContext.Provider value={{ onChange: onChangeMock }}>
            <ThemeProvider>
              <RadioGroup.Button>Radio 1</RadioGroup.Button>
            </ThemeProvider>
          </RadioGroupContext.Provider>,
        );

        const radio = getByText(/^Radio/);

        fireEvent.press(radio);

        expect(onChangeMock).toHaveBeenCalled();
      });

      it('value should be as prop value', () => {
        const onChangeMock = jest.fn();
        const { getByText } = render(
          <RadioGroupContext.Provider value={{ onChange: onChangeMock }}>
            <ThemeProvider>
              <RadioGroup.Button value="radio1">Radio 1</RadioGroup.Button>
            </ThemeProvider>
          </RadioGroupContext.Provider>,
        );

        const radio = getByText(/^Radio/);

        fireEvent.press(radio);

        expect(onChangeMock).toHaveBeenCalledWith({
          value: 'radio1',
        });
      });

      it('value should be children if prop value is ommited', () => {
        const onChangeMock = jest.fn();
        const { getByText } = render(
          <RadioGroupContext.Provider value={{ onChange: onChangeMock }}>
            <ThemeProvider>
              <RadioGroup.Button>Radio 1</RadioGroup.Button>
            </ThemeProvider>
          </RadioGroupContext.Provider>,
        );

        const radio = getByText(/^Radio/);

        fireEvent.press(radio);

        expect(onChangeMock).toHaveBeenCalledWith({
          value: 'Radio 1',
        });
      });
    });
  });
});
