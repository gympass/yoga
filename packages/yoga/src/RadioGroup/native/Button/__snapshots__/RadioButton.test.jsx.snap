// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioGroup /> <RadioGroup.Button /> Snapshots should match snapshot with default radio button 1`] = `
<View
  accessible={true}
  checked={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "borderRadius": 9999,
        "fontSize": 14,
        "height": 40,
        "justifyContent": "center",
        "paddingLeft": 24,
        "paddingRight": 24,
      },
    ]
  }
>
  <Text
    checked={false}
    style={
      Array [
        Object {
          "color": "#231B22",
          "fontFamily": "Rubik",
        },
      ]
    }
  >
    Radio 1
  </Text>
</View>
`;

exports[`<RadioGroup /> <RadioGroup.Button /> Snapshots should match snapshot with small radio button 1`] = `
<View
  accessible={true}
  checked={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  small={true}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "borderRadius": 9999,
        "fontSize": 14,
        "height": 32,
        "justifyContent": "center",
        "paddingLeft": 24,
        "paddingRight": 24,
      },
    ]
  }
>
  <Text
    checked={false}
    style={
      Array [
        Object {
          "color": "#231B22",
          "fontFamily": "Rubik",
        },
      ]
    }
  >
    Radio 1
  </Text>
</View>
`;
